pkg/server/message.go

```diff
@@ -236,18 +233,117 @@ func (t *TranslateServiceServer) UpdateMessages(
 		return nil, status.Errorf(codes.Internal, "")
 	}
 
-	// If updating the original messages and populateTranslations flag is true, update all translated messages as well.
-	if params.messages.Original && params.populateTranslations {
-		if err := t.populateTranslatedMessages(ctx, params.serviceID, params.messages, msgs); err != nil {
+	if params.messages.Original && len(allMessages) > 1 {
+		// TODO: optimize performance when populateTranslations param is true, currently saveMessages() will be called twice.
+
+		// find original messages where text has been altered then translate & update associated messages for all translations.
+		if err := t.updateAlteredMessages(ctx, params.serviceID, allMessages, params.messages); err != nil {
 			return nil, err
 		}
+
+		// if populateTranslations flag is true, update all translated messages as well.
+		if params.populateTranslations {
+			if err := t.populateTranslatedMessages(ctx, params.serviceID, params.messages, allMessages); err != nil {
+				return nil, err
+			}
+		}
 	}
 
 	return messagesToProto(params.messages), nil
 }
 
 // helpers
 
+// updateAlteredMessages finds original messages where text has been altered,
+// translates & updates associated messages for all translations.
+func (t *TranslateServiceServer) updateAlteredMessages(
+	ctx context.Context,
+	serviceID uuid.UUID,
+	allMessages []model.Messages,
+	newOriginalMessages *model.Messages,
+) error {
+	// return if only one language is present
+	if len(allMessages) == 1 {
+		return nil
+	}
+
+	var originalMessages *model.Messages
+
+	for i := range allMessages {
+		if allMessages[i].Original {
+			originalMessages = &allMessages[i]
+			break
+		}
+	}
+
+	// return if messages don't contain original language
+	if originalMessages == nil {
+		return nil
+	}
+
+	// create lookup for original messages
+	originalMessagesLookup := make(map[string]*model.Message, len(originalMessages.Messages))
+
+	for i := range originalMessages.Messages {
+		originalMessagesLookup[originalMessages.Messages[i].ID] = &originalMessages.Messages[i]
+	}
+
+	// find altered original messages
+	alteredMessages := model.Messages{
+		Language: newOriginalMessages.Language,
+		Original: true,
```

Comment:
this is on the question mark, but how about reserving memory for the length of original messages `len(originalMessages.Messages)`, instead of letting it grow automatically?

Reply:
What if there are thousands of messages, but only one of them is altered?

Reply:
Depends on the frontend, if it calls `updateMessages` on every translated messages, then yes, but if it is called only once, after the translator, is finished translating some batch, then that would be another story

Reply:
Why `Original: true,`?

```suggestion
```

Reply:
umm, just added it for additional clarification that those are original messages.